### PR TITLE
Rewrite concat to start producers iteratively

### LINKED ISSUE #3140: Flatten with .Concat stacks overflow bug when tons of completed producer need to be processed
We ran into a bug, when we have tons of completed producer in the SignalProducer pipeline inside a flatten with .Concat, as it seems .Concat was implemented with recursive calls for the next value, the stack will overflow then

Here is the simple code to reproduce it

```Swift
  func testConcatOverflow() {
    var values: [Int] = []
    for n in 0..<10000 {
      values.append(n)
    }

    SignalProducer<Int, NSError>(values: values)
      .flatMap(.Concat) { n -> SignalProducer<Int, NSError> in
        if n == 0 {
          return SignalProducer<Int, NSError>(value: n)
            .delay(0.1, onScheduler: QueueScheduler.mainQueueScheduler)
        }
        return SignalProducer<Int, NSError>(value: n)
      }
      .startWithNext { n in
        print("Num \(n)")
      }
    let exp = expectationWithDescription("")
    waitForExpectationsWithTimeout(100, handler: nil)
  }
```

Surely the bug can be workarounded via adding a delay to any queue for the SignalProducer, but I wonder why not implement .Concat with loop instead of recursion?

It's kind of surprising behavior to me, also very hard to debug, took me whole day.

I will try to fix the problem with a loop implementation of .Concat and see what you guys think about it.

### PR DESCRIPTION
**Pull request by mdiep**
Recursively starting the inner producers can blow the stack if they complete immediately.

Fixes #3140.

Supersedes #3141.

### CODE REVIEW

**Comment by andersio** [comment]
It feels like the `ConcatState` would get meaningful method names if we take out the recursive/loop tracking aspect.

**Comment by mdiep** [comment]
That seems to me like an integral part of the improvement here. 😕

File: ReactiveCocoa/Swift/Flatten.swift [hunk 1]
```diff
@@ -341,25 +341,50 @@ extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Err
...
+				producer.startWithSignal { signal, inner in
+					let handle = disposable?.add(inner) ?? nil
+
+					signal.observe { event in
+						switch event {
+						case .completed, .interrupted:
+							handle?.remove()
 
-		return self.observe { event in
+							if state.modify({ $0.didTerminate() }) {
```
Comment (andersio) [inline ReactiveCocoa/Swift/Flatten.swift]: We might need some clarity at the call site.

File: ReactiveCocoa/Swift/Flatten.swift [hunk 2]
```diff
@@ -341,25 +341,50 @@ extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Err
...
 		}
 	}
-
+	
 	fileprivate func observeConcat(_ observer: Observer<Value.Value, Error>, _ disposable: CompositeDisposable? = nil) -> Disposable? {
-		let state = ConcatState(observer: observer, disposable: disposable)
+		let state = Atomic(ConcatState<Value.Value, Error>())
+		
+		func startNextIfNeeded() {
+			while let producer = state.modify({ $0.willStart() }) {
```
Comment (andersio) [inline ReactiveCocoa/Swift/Flatten.swift]: How about `dequeue` and `acknowledgeStarted`?

**Comment by mdiep** [comment]
I went with `dequeue`, but got rid of the other methods and just set the properties directly. The names were one thing I didn't like about this, and I think this is an improvement.

File: ReactiveCocoa/Swift/Flatten.swift [hunk 3]
```diff
@@ -341,25 +341,52 @@ extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Err
...
+
+					signal.observe { event in
+						switch event {
+						case .completed, .interrupted:
+							handle?.remove()
+							
+							state.modify { $0.active = nil }
 
-		return self.observe { event in
+							if !state.value.isStarting {
```
Comment (andersio) [inline ReactiveCocoa/Swift/Flatten.swift]: You can return `isStarting` from the action above to avoid acquiring the lock twice.

**Comment by mdiep** [comment]
Closed in favor of https://github.com/ReactiveCocoa/ReactiveSwift/issues/5.

**Pull request closed without merging**